src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -0,0 +1,199 @@
+            ADP.CheckArgumentNull(factory, nameof(factory));
+
+            _registeredFactories[providerInvariantName] = new ProviderRegistration(factory.GetType().AssemblyQualifiedName, factory);
+        }
+        
+        public static bool UnregisterFactory(string providerInvariantName)
+        {
+            return !string.IsNullOrWhiteSpace(providerInvariantName) && _registeredFactories.TryRemove(providerInvariantName, out ProviderRegistration providerRegistration);
+        }
+        
+        private static DbProviderFactory GetFactory(string providerInvariantName, bool throwOnError)
+        {
+            if (throwOnError)
+            {
+                ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+            }
+            else
+            {
+                if (string.IsNullOrWhiteSpace(providerInvariantName))
+                {
+                    return null;
+                }
+            }
+            bool wasRegistered = _registeredFactories.TryGetValue(providerInvariantName, out ProviderRegistration registration);
+            if (!wasRegistered)
+            {
+                return throwOnError ? throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_InvariantNameNotFound, providerInvariantName)) : (DbProviderFactory)null;
+            }
+            DbProviderFactory toReturn = registration.FactoryInstance;
+            if (toReturn == null)
+            {
+                // Deferred registration, do checks now on the type specified and register instance in storage.
+                // Even in the case of throwOnError being false, this will throw when an exception occurs checking the registered type as the user has to be notified the 
+                // registration is invalid, even though the registration is there.
+                Type providerType = Type.GetType(registration.FactoryTypeAssemblyQualifiedName);
+                if (null == providerType)
+                {
+                    throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_FactoryNotLoadable, registration.FactoryTypeAssemblyQualifiedName));
+                }
+                toReturn = GetFactoryInstance(providerType);
+                RegisterFactory(providerInvariantName, toReturn);
+            }
+            return toReturn;
+        }
+       
+        private static DbProviderFactory GetFactoryInstance(Type providerFactoryClass)
+        {
+            ADP.CheckArgumentNull(providerFactoryClass, nameof(providerFactoryClass));
+            if (!providerFactoryClass.IsSubclassOf(typeof(DbProviderFactory)))
+            {
+                throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_NotAFactoryType, providerFactoryClass.FullName));
+            }
+
+            System.Reflection.FieldInfo providerInstance = providerFactoryClass.GetField(Instance, System.Reflection.BindingFlags.DeclaredOnly | System.Reflection.BindingFlags.Public | 
```

Comment:
Nit: using?

Reply:
Added using, cleaned up the BindingFlags statements too.